File: tools/verify-native-linkage.py
#!/usr/bin/env python3
"""Hold the Linux candidate to its documented system-runtime linkage boundary."""
import pathlib
import re
import subprocess
import sys

module = pathlib.Path(sys.argv[1]) if len(sys.argv) == 2 else pathlib.Path(__file__).resolve().parent.parent / 'modules/kumwe_engine.so'
if len(sys.argv) > 2 or not module.is_file():
    raise SystemExit('Usage: verify-native-linkage.py [BUILT_MODULE]')
dynamic = subprocess.check_output(['readelf', '-d', str(module)], text=True)
needed = set(re.findall(r'Shared library: \[([^]]+)\]', dynamic))
allowed = {'libstdc++.so.6', 'libm.so.6', 'libgcc_s.so.1', 'libc.so.6', 'libpthread.so.0', 'librt.so.1'}
if not needed or needed - allowed:
    raise SystemExit('Unexpected extension shared-library dependencies: ' + ', '.join(sorted(needed - allowed)))
exports = subprocess.check_output(['nm', '-D', '--defined-only', str(module)], text=True)
if re.search(r'\bpcre2_[A-Za-z0-9_]+', exports):
    raise SystemExit('Vendored PCRE2 internals must remain hidden inside the native module.')
print('Native module uses only documented system runtimes: ' + ', '.join(sorted(needed)))

File: tools/verify-native-linkage.php
#!/usr/bin/env php
<?php
declare(strict_types=1);
namespace Kumwe\Tools\NativeLinkage;
require_once __DIR__ . '/diagnostic-runtime.php';

use Kumwe\Tools\DiagnosticRuntime\Invalid;
use function Kumwe\Tools\DiagnosticRuntime\run;

try {
    $module = $argv[1] ?? __DIR__ . '/../modules/kumwe_engine.so';
    if ($argc > 2 || !is_file($module)) {
        throw new Invalid('Usage: verify-native-linkage.php [BUILT_MODULE]');
    }
    $dynamic = run('readelf', '-d', $module);
    preg_match_all('/Shared library: \[([^]]+)\]/', $dynamic, $matches);
    $needed = array_unique($matches[1]);
    sort($needed, SORT_STRING);
    $allowed = ['libstdc++.so.6', 'libm.so.6', 'libgcc_s.so.1', 'libc.so.6', 'libpthread.so.0', 'librt.so.1'];
    $unexpected = array_diff($needed, $allowed);
    if (!$needed || $unexpected) {
        throw new Invalid('Unexpected extension shared-library dependencies: ' . implode(', ', $unexpected));
    }
    $exports = run('nm', '-D', '--defined-only', $module);
    if (preg_match('/\bpcre2_[A-Za-z0-9_]+/', $exports)) {
        throw new Invalid('Vendored PCRE2 internals must remain hidden inside the native module.');
    }
    echo 'Native module uses only documented system runtimes: ' . implode(', ', $needed) . "\n";
} catch (\Throwable $error) {
    fwrite(STDERR, $error->getMessage() . "\n");
    exit(1);
}
